additional Oracle12c Identity Columns

ShintaroTamaki opened the pull request:
実装してみました、ご確認よろしくお願いいたします。

ShintaroTamaki committed:
additional Oracle12c Identity Columns

jflute commented:
@ShintaroTamaki おおぉ、ありがとうございます。
ちょっと見るの時間掛かるかもですが、時間見つけて見ていきますね。

ちなみに、そちらの環境では、コンパイルやビルドは大丈夫でした？
Eclipseのバージョンはいくつでしょうか？

jflute commented:
Issueへのリンク: 
https://github.com/dbflute-session/erflute/issues/50

ShintaroTamaki commented:
@jflute 
ありがとうございます、よろしくおねがいします。

こちら、下記環境で修正しております。
```
Windows10 2021H2最新版パッチ反映済み

Eclipse IDE for Enterprise Java and Web Developers (includes Incubating components)
Version: 2021-12 (4.22.0)
Build id: 20211202-1639
```
あいにく、Macでテスト可能な環境がないためMacの確認ができておりません。

テストは、ホームページのままで実施できました。
http://dbflute.seasar.org/ja/manual/function/helper/erflute/index.html#testway

jflute commented:
@ShintaroTamaki おお、Eclipse-2021-12 で動いたんですね。それは良かったです。
であれば自分の方でもプラグインとか整えれば動くはずなので、やってみますね

ShintaroTamaki commented:
@jflute
あと、Javaバージョンです。
```
openjdk 17.0.2 2022-01-18
OpenJDK Runtime Environment Temurin-17.0.2+8 (build 17.0.2+8)
OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (build 17.0.2+8, mixed mode)
```

zoosm3 commented on resources/SqlType.tsv:
```diff
@@ -1,136 +1,136 @@
-sqlTypeId	javaclass	need args	full text indexable	StandardSQL	DB2	H2	HSQLDB	MySQL	Oracle	PostgreSQL	SQLite	SQLServer	SQLServer 2008	MSAccess
```
改行コードが変わってしまっているようです。
追加内容は `Oracle12c` だけに見えますので、理想はその部分だけ差分が出ることだと思います。

jflute replied:
分析ありがとうございますー。

ShintaroTamaki replied:
改行コードをLFに修正しました。

zoosm3 commented on src/org/dbflute/erflute/db/impl/oracle12c/Oracle12cDDLCreator.java:
```diff
@@ -0,0 +1,67 @@
+            ddl.append(" NOT NULL");
+        }
+        if (normalColumn.isUniqueKey()) {
+            if (!Check.isEmpty(normalColumn.getUniqueKeyName())) {
```
> jfluteさん

`// #thinking jflute all right? (e.g. bad in MySQL) (2020/05/16)`
https://docs.oracle.com/cd/E82638_01/sqlrf/constraint.html#GUID-1055EA97-BA6F-4764-A15F-1024FD5B6DFE
Oracleに関しては問題なさそうです

jflute replied:
調査ありがとうございます。
コメントのそのまんまリンクを貼っておきたいですね。

ShintaroTamaki replied:
コメントを入れました。
https://github.com/ShintaroTamaki/erflute/blob/feature/additional_oracle12c_identity_column/src/org/dbflute/erflute/db/impl/oracleidentity/OracleIdentityDDLCreator.java#L55-L56

zoosm3 commented on src/org/dbflute/erflute/db/impl/oracle12c/Oracle12cDDLCreator.java:
```diff
@@ -0,0 +1,67 @@
+                ddl.append(normalColumn.getUniqueKeyName());
+            }
+            ddl.append(" UNIQUE");
+        }
```
https://docs.oracle.com/cd/F44649_01/sqlreferencefornosql/identity-column.html
実装上 `BY DEFAULT` で固定されていますが、ERFluteで`GENERATED ALWAYS` をサポートしなくても日々のお仕事に影響はないでしょうか。
Oracle 12gで開発したことがないですが、アイデンティティ列の値を常に使いたいユースケースは少なそうですし、一旦は対応しなくていいのかもしれませんが。

jflute replied:
なるほど、他のDBでもどうなってるか？ERFluteとして方向性がブレていなければですね。

https://github.com/dbflute-session/erflute/blob/develop/src/org/dbflute/erflute/db/impl/db2/DB2DDLCreator.java#L24
```
        if (normalColumn.isAutoIncrement()) {
            ddl.append(" GENERATED ALWAYS AS IDENTITY ");
```

DB2は、ALWAYSがデフォルトになってますね。。。(ってか選べない)
ALWAYSは不便なこともあるんですけどね。


@ShintaroTamaki たまきさん、実際に業務で利用したいのは、BY DEFAULT でしょうか？

ShintaroTamaki replied:
こちらの記事を確認し、DBFluteのreplace_schemaも利用するため、ALWAYS は採用しませんでした。
https://oracle-base.com/articles/12c/identity-columns-in-oracle-12cr1

で、主キーがNullになることはないと考え、BY DEFAULTを採用しております。

jflute replied:
@ShintaroTamaki ありがとうございます。
そうなんですよね。ALWAYSは融通が効かなくて...
DBFluteだったらどのみちEntityにIDをsetしてinsertしてもIdentityを強制しますし。

どっちかっていうとDB2の方を直したい勢いですね（＾＾。

ShintaroTamaki replied:
DB2を修正して正常な動作になるか不明なため、対応しておりません。

jflute replied:
りょうかいです。Oracle は BY DEFAULT でOKです。
そして、DB2は放置でOKです。

zoosm3 commented on src/org/dbflute/erflute/db/impl/oracle12c/reserved_word.properties:
```diff
@@ -0,0 +1,469 @@
+ACCESS
```
[コメント]予約後一覧の正しさは確認していません

zoosm3 commented on src/org/dbflute/erflute/db/DBManagerFactory.java:
```diff
@@ -32,6 +33,7 @@
         new AccessDBManager();
         new MySQLDBManager();
         new OracleDBManager();
```
`new Oracle12cDBManager()`

- OracleDBManagerが <= Oracle 11g まで
- Oracle12cDBManagerが => Oracle 12c 以上
に対応する `DBManager` のクラスってことですよね。
対応するOracleのバージョンが分かりにくいと思いました。
名前の改善など考えられるでしょうか。

jflute replied:
そうですね。ちょっとバージョン対応をどうするかはERFlute自身でまだ整理されていないことですね。
基本的には、挙動が変わったわけじゃないので、OracleDBManagerに入れ込んじゃっても良いかなと思っています。

Oracle12cは2013年リリースのようでずいぶん前ですし、古いOracleを使ってる人はチェックを付ければ良いだけなので。
(チェックを付けたらDDLが落ちるので気付きますし)

ShintaroTamaki replied:
OracleDBManagerのAutoIncrementにチェックした場合、DDL Exportでシーケンスとトリガーが作成されます。
なので、その機能を生かすために、分離したほうが良いと思い、分離してます。

zoosm3 replied:
そうですね。開発環境など最新化してリリースするのであれば、告知の上ERFluteのサポートするOracleバージョンを12以上にするのであれば、OracleDBManagerを上書きしてもいいと思いました。

jflute replied:
@ShintaroTamaki たまきさん、ありがとうございます。なんと！そういうことなんですね。
既存がAutoIncrementをSequence+Trigger方式で実現しちゃってるから挙動が変わってしまうわけですね。
ぐぬぬぬ。ちょっと考えますね。

でも、もうEclipseのバージョンも上がるリリースになるので「最新Oracleに合わせちゃう」でも良いかなとは思ってるんですよね。
オプションで以前の方式に変えられるってのは理想みたいな。

ShintaroTamaki replied:
クラス名・パッケージ名を Oracle12c → OracleIdentity に変更しました。

jflute replied:
@ShintaroTamaki たまきさん、ありがとうございます。
ちょっともう決めなんですけど...

**「OracleのSequence/Trigger方式のIdentityはサポートせず、Oracle12cのIdentityをメインにする」**

という風にしてしまいたいと思います。

理由は...
o もうOracle12cは2013年リリースと9年も経っている
o OracleでSequence/Trigger方式でIdentity的に扱ってるのを(個人的に)聞いたことがない
o OracleでERFluteを使っている事例を(個人的に)今まで聞いたことがない、今回初めて
o いざ必要になれば、オプションで以前の方式をサポートするという風にもできる
(o Eclipseのバージョンも新しいものでリリースするので区切りも良い)

ということでお手数なんですが、OracleDBManagerに直接実装を入れ込むように実装できますでしょうか？

jflute replied:
[memo] DBFlute Slackでの議論において。

```
A. 超理想: DBのバージョンを入れられるようにして、バージョンごとに挙動が変わるようにする
B. 理想: AutoIncrementのところで実現方法を良い感じに選択できるUIで切り替える
C. 現実: Advanced Settingsで切り替えられるようにする
D. 超現実: いったん機能オフにする

※どっちをデフォルトにするか？ってのは再考してもいいかも。
Identityをオプションにするか？DefaultSequenceをオプションにするか？
```

Bでチャレンジしてもらうことに。

jflute reviewed (commented):
取り急ぎ、zoosm3さんのコメントへの反応させて頂きました。
実際にどう修正していくか？など、また後で(週末かも!?)じっくり書きます。

zoosm3 reviewed (commented):
気になる点があるのでコメントしましたが、質問と感想文なので修正はお任せいたします。
Oracle 12gのドキュメントを見て、DBManagerで生成しているIdentity Columnsの構文があっていることを確認しました。
テストをして期待通りカラムが生成されれば特に問題なさそうです。

https://github.com/dbflute-session/erflute/issues/3
で上がっていましたが、環境構築時のことをREADMEに書き起こしていただけると便利だと思いました。
私も環境構築を試してみようと思います。

ShintaroTamaki committed:
修正内容
・GENERATED BY DEFAULT AS IDENTITYがNOT NULLの前でないとエラーになったため修正
・パッケージ、クラス名 Oracle12c → OracleIdentity
・resources/SqlType.tsv 改行コードをLFに変更
・レビュー指摘事項反映

ShintaroTamaki commented:
> #3 で上がっていましたが、環境構築時のことをREADMEに書き起こしていただけると便利だと思いました。 私も環境構築を試してみようと思います。


http://dbflute.seasar.org/ja/manual/function/helper/erflute/index.html#forcommitter

ShintaroTamaki commented:
修正しました、確認お願いいたします。

あらたな修正点は下記となります。
・ GENERATED BY DEFAULT AS IDENTITYがNOT NULLの前でないとエラーになったため修正

いただいておりました指摘事項に関しては、レスをつけております。

jflute commented:
memo: 流れをわかりやすくするためのコメント。

現状フォーカスしてる課題は、以下のdiscussionを参照。
https://github.com/dbflute-session/erflute/pull/51#discussion_r815780108

```
A. 超理想: DBのバージョンを入れられるようにして、バージョンごとに挙動が変わるようにする
B. 理想: AutoIncrementのところで実現方法を良い感じに選択できるUIで切り替える
C. 現実: Advanced Settingsで切り替えられるようにする
D. 超現実: いったん機能オフにする

※どっちをデフォルトにするか？ってのは再考してもいいかも。
Identityをオプションにするか？DefaultSequenceをオプションにするか？

↑Bでチャレンジ
```

zoosm3 commented:
@jflute @ShintaroTamaki 
こちら1年と少し経過しているようですが、お手伝いできることなどありますか？
せっかくここまで来たのでリリースして使えるようにできるのが良いと思いまして。

jflute commented:
> @jflute @ShintaroTamaki
> こちら1年と少し経過しているようですが、お手伝いできることなどありますか？
> せっかくここまで来たのでリリースして使えるようにできるのが良いと思いまして。

@zoosm3 一年越しの返信で申し訳ないですが、Yuさんお気遣いありがとうございます。
Tamakiさんもちょっとお忙しそうですから、なかなか難しそうですね。

いったん別ブランチにマージして、せめて以下の現実ラインに持っていけたらよいのかもですね。
```
C. 現実: Advanced Settingsで切り替えられるようにする
```

なんにせよ、現状分析から始めないと...

zoosm3 commented:
@jflute こちらこそ1年越しのコメントありがとうございます、。一旦様子見しますね
